Added Epoch Support

GabrielNotman opened the pull request:
Updated 23/11/15
Removed: [Added full H12 AM/PM support for setting the hours value (also for querying).]

Previously the RTC_MODE2_CTRL_CLKREP bit was not being set correctly (inverted).

Added the ability to get or set the time and date using either a Epoch or Y2K timestamp.
Epoch timestamps are clamped to Y2K lower bound.

Removed: [A few other modifications including renaming the function pointer type to avoid possible
conflicts and moving the __time24 into the class.]

agdl commented:
@GabrielNotman  Sorry but I removed the H12/H24 time representation because was asked in an Arduino issue, so now this PR has conflicts

agdl commented:
What can we do about this? Having two begin function prototypes?

agdl commented:
@GabrielNotman can you please join the discussion in the referenced issue?

GabrielNotman commented:
I'm fine with removing the 12hr mode support. It complicates things and creates potential confusion.
It is also trivial for anyone to convert if they wish to.

One note: I'm fairly certain that the RTC_MODE2_CTRL_CLKREP bit is asserted for 12h mode. You can test by setting the hours to 13, in 12 hour mode it will adjust that to 1pm.

agdl commented:
Nope if you set 13 you have 13! However have you seen this commit?
https://github.com/arduino-libraries/RTCZero/commit/d67cdb15969c09b6df45a90c5a2aee245a03c09a

It enables the standby mode and shows how to use it!

GabrielNotman commented:
The main reason I needed the alarm functionality was for a wake interrupt.
That is a nice example demonstrating this.

May I suggest adding a conditional to check if the SerialUSB is present,
and only go to sleep if it is not. This will remove any issues people may
have if they are programming it via the Native USB port.

I'm a bit confused about that mode bit. I will try and check it tonight
when I
have time.

One thing, I believe if you set out of range values, it will correct them.
However, it may only do so after the next clock cycle (1 second).

On 18 November 2015 at 11:33, Arturo Guadalupi <notifications@github.com>
wrote:

> Nope if you set 13 you have 13! However have you seen this commit?
> d67cdb1
> <https://github.com/arduino-libraries/RTCZero/commit/d67cdb15969c09b6df45a90c5a2aee245a03c09a>
>
> It enables the standby mode and shows how to use it!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arduino-libraries/RTCZero/pull/3#issuecomment-157673386>
> .
>



-- 

Regards,


Gabriel Notman

Software Developer at SODAQ


If our environment could speak, what would it say?

Start connecting devices everywhere.


www.sodaq.com

agdl commented:
Feel free to improve the examples and any related things as you wish ;)

GabrielNotman commented:
Do you need me to create a new PR using the current code, in order to
resolve
the conflicts with the existing PR?

On 18 November 2015 at 12:07, Arturo Guadalupi <notifications@github.com>
wrote:

> Feel free to improve the examples and any related things as you wish ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arduino-libraries/RTCZero/pull/3#issuecomment-157680393>
> .
>



-- 

Regards,


Gabriel Notman

Software Developer at SODAQ


If our environment could speak, what would it say?

Start connecting devices everywhere.


www.sodaq.com

agdl commented:
I don't know it depends on how you want to resolve them!

GabrielNotman commented:
I've removed the conflicts and you should be able to merge it now.

A few notes:
I have removed the assertion of the CLKREP bit of the CTRL register.
It should be 0 for 24 hour mode. See page 245 of the SAMD21 datasheet.
You can test this using the SimpleRTC and an initial time of 12:59.

Should the standbyMode() method be part of this clock class?

GabrielNotman commented:
Any update on when this pull request might be processed?
